IPv6 is generally broken for the S-CSCF sproutlet

If the sprout_hostname is an IPv6 address then the S-CSCF plugin fails to load as the BGCF URI is invalid 

Debug scscfsproutlet.cpp:99:   BGCF URI           = sip:bgcf.[::1]:5054;transport=TCP
Error scscfsproutlet.cpp:125: Invalid BGCF URI sip:bgcf.[::1]:5054;transport=TCP

A workaround to the above is to make the sprout_hostname a hostname. This then crashes on the S-CSCF node URI:

Debug scscfsproutlet.cpp:95: Creating S-CSCF Sproutlet
Debug scscfsproutlet.cpp:96:   S-CSCF cluster URI = sip:sprout.em4.cw-ngv.com:5054;transport=TCP
Debug scscfsproutlet.cpp:97:   S-CSCF node URI    = sip:::1:5054
Debug scscfsproutlet.cpp:98:   I-CSCF URI         = sip:sprout.em4.cw-ngv.com:5052;transport=TCP
Debug scscfsproutlet.cpp:99:   BGCF URI           = sip:bgcf.sprout.em4.cw-ngv.com:5054;transport=TCP
Error scscfsproutlet.cpp:117: Invalid S-CSCF node URI sip:::1:5054

This is built from the local_ip value in /etc/clearwater/config. We can't just wrap this in square brackets, as pj_getaddrinfo fails to parse this, e.g.:

Error stack.cpp:220: Failed to decode IP address [::1] (gethostbyname() has returned error (PJ_ERESOLVE))

Raising the priority as there's not a simple workaround

--- a/sprout/icscfplugin.cpp
+++ b/sprout/icscfplugin.cpp
@@ -86,7 +86,7 @@ bool ICSCFPlugin::load(struct options& opt, std::list<Sproutlet*>& sproutlets)
     // Determine the S-CSCF and hence BGCF URIs.
     std::string scscf_cluster_uri = std::string(stack_data.scscf_uri.ptr,
                                                 stack_data.scscf_uri.slen);
-    std::string bgcf_uri = "sip:bgcf." + scscf_cluster_uri.substr(4);
+    std::string bgcf_uri = "sip:bgcf@" + scscf_cluster_uri.substr(4);
 
     // Create the S-CSCF selector.
     _scscf_selector = new SCSCFSelector();

--- a/sprout/scscfplugin.cpp
+++ b/sprout/scscfplugin.cpp
@@ -89,7 +89,7 @@ bool SCSCFPlugin::load(struct options& opt, std::list<Sproutlet*>& sproutlets)
     }
 
     std::string scscf_node_uri = "sip:" + node_ip + ":" + std::to_string(opt.scscf_port);
-    std::string bgcf_uri = "sip:bgcf." + scscf_cluster_uri.substr(4);
+    std::string bgcf_uri = "sip:bgcf@" + scscf_cluster_uri.substr(4);
     std::string icscf_uri;
     if (opt.icscf_enabled)
     {
@@ -106,8 +106,8 @@ bool SCSCFPlugin::load(struct options& opt, std::list<Sproutlet*>& sproutlets)
       }
       else
       {
-        // No port number, so best we can do is strap icscf. on the front.
-        icscf_uri = "sip:icscf." + scscf_cluster_uri.substr(4);
+        // No port number, so best we can do is strap 'icscf@' on the front.
+        icscf_uri = "sip:icscf@" + scscf_cluster_uri.substr(4);
       }
     }
     else

--- a/sprout/scscfplugin.cpp
+++ b/sprout/scscfplugin.cpp
@@ -40,6 +40,7 @@
  */
 
 #include "cfgoptions.h"
+#include "ipv6utils.h"
 #include "sproutletplugin.h"
 #include "scscfsproutlet.h"
 
@@ -80,10 +81,14 @@ bool SCSCFPlugin::load(struct options& opt, std::list<Sproutlet*>& sproutlets)
     // Determine the S-CSCF, BGCF and I-CSCF URIs.
     std::string scscf_cluster_uri = std::string(stack_data.scscf_uri.ptr,
                                                 stack_data.scscf_uri.slen);
-    std::string scscf_node_uri = "sip:" +
-                                 std::string(stack_data.local_host.ptr,
-                                             stack_data.local_host.slen) +
-                                 ":" + std::to_string(opt.scscf_port);
+    std::string node_ip(stack_data.local_host.ptr, stack_data.local_host.slen);
+    
+    if (is_ipv6(node_ip))
+    {
+      node_ip = "[" + node_ip + "]";
+    }
+
+    std::string scscf_node_uri = "sip:" + node_ip + ":" + std::to_string(opt.scscf_port);
     std::string bgcf_uri = "sip:bgcf." + scscf_cluster_uri.substr(4);
     std::string icscf_uri;
     if (opt.icscf_enabled)

--- a/sprout/ut/scscf_test.cpp
+++ b/sprout/ut/scscf_test.cpp
@@ -352,7 +352,7 @@ public:
     _scscf_sproutlet = new SCSCFSproutlet("sip:homedomain:5058",
                                           "sip:127.0.0.1:5058",
                                           "",
-                                          "sip:bgcf.homedomain:5058",
+                                          "sip:bgcf@homedomain:5058",
                                           5058,
                                           _store,
                                           NULL,

